File: PQC/BouncyCastle/MyBcPQC/TestMyBcPQC/TestBcPqcSignatur/Stateless/Test-SLH-DSA/BobSLHDSA.cs

using Org.BouncyCastle.Crypto.Generators;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;
using Org.BouncyCastle.Security;

namespace michele.natale.TestBcPqcs;

using BcPqcs;
using Services;

public class BobSLHDSA : IDisposable
{
  public byte[] PubKey { get; private set; } = null!;
  public bool IsDisposed { get; private set; } = true;
  public SlhDsaKeyPairInfo Info { get; private set; } = null!;
  public SlhDsaParameters Parameter { get; private set; } = null!;

  private SecureRandom Rand = new();


  public BobSLHDSA()
    : this(ToRngParameter())
  {
  }

  public BobSLHDSA(SlhDsaParameters parameter)
  {
    this.Parameter = parameter;

    var generator = new SlhDsaKeyGenerationParameters(
      this.Rand, this.Parameter);

    var keypair_generator = new SlhDsaKeyPairGenerator();
    keypair_generator.Init(generator);

    var keypair = keypair_generator.GenerateKeyPair();
    this.PubKey = ((SlhDsaPublicKeyParameters)keypair.Public).GetEncoded();
    var privkey = ((SlhDsaPrivateKeyParameters)keypair.Private).GetEncoded();
    this.Info = new SlhDsaKeyPairInfo(this.PubKey, privkey, parameter);

    Array.Clear(privkey);
    this.IsDisposed = false;
  }

  public void Clear()
  {
    if (this.IsDisposed) return;

    if (this.PubKey is not null)
      Array.Clear(this.PubKey);

    this.Rand = null!;
    this.Info = null!;
    this.PubKey = null!;
    this.Parameter = null!;

  }

  public SlhDsaPublicKeyParameters ToPubKey() =>
    SlhDsaPublicKeyParameters.FromEncoding(
      this.Parameter, this.PubKey);


  public (byte[] Sign, byte[] PubKey) Sign(ReadOnlySpan<byte> message)
  {
    var msg = message.ToArray();

    //SLH-DSA Parameter 
    var parameter = this.Info.ToParameter();

    //Sign Message-Data
    var signer = new SlhDsaSigner(parameter, true);
    var privkey = SlhDsaPrivateKeyParameters
      .FromEncoding(parameter, this.Info.ToPrivKey().ToBytes());
    signer.Init(true, privkey);
    signer.BlockUpdate(msg, 0, msg.Length);

    return (signer.GenerateSignature(), this.PubKey);
  }

  public static bool Verify(
    ReadOnlySpan<byte> pubkeyalice,
    ReadOnlySpan<byte> signature,
    ReadOnlySpan<byte> message,
    SlhDsaParameters parameter)
  {
    var msg = message.ToArray();

    //Verify Signature
    var signer = new SlhDsaSigner(parameter, true);
    var pubkey = SlhDsaPublicKeyParameters.FromEncoding(
      parameter, pubkeyalice.ToArray());
    signer.Init(false, pubkey);
    signer.BlockUpdate(msg, 0, msg.Length);
    return signer.VerifySignature(signature.ToArray());
  }

  private static SlhDsaParameters ToRngParameter()
  {
    var rand = new SecureRandom();
    var parameters = BcPqcServices.ToSLHDsaParameters();
    var idx = rand.Next(parameters.Length);
    return parameters[idx];
  }

  public static SlhDsaPublicKeyParameters ToPubKey(
    byte[] pubkey, SlhDsaParameters parameter) =>
      SlhDsaPublicKeyParameters.FromEncoding(parameter, pubkey);

  protected virtual void Dispose(bool disposing)
  {
    if (!this.IsDisposed)
    {
      if (disposing)
      {
      }
      this.IsDisposed = true;
    }
  }

  ~BobSLHDSA() => Dispose(false);

  public void Dispose()
  {
    Dispose(true);
    GC.SuppressFinalize(this);
  }
}

File: PQC/BouncyCastle/MyBcPQC/MyBcPQC/TestMyBcPQCVb/TestBcPqcSignatur/Stateless/Test-SLH-DSA/BobSLHDSA.vb
Option Strict On
Option Explicit On



Imports michele.natale.BcPqcs
Imports michele.natale.Services
Imports Org.BouncyCastle.Security
Imports Org.BouncyCastle.Crypto.Signers
Imports Org.BouncyCastle.Crypto.Generators
Imports Org.BouncyCastle.Crypto.Parameters


Namespace michele.natale.TestBcPqcs

  Public Class BobSLHDSA
    Implements IDisposable

    Private MPubKey As Byte() = Array.Empty(Of Byte)
    Public Property PubKey() As Byte()
      Get
        Return Me.MPubKey
      End Get
      Private Set
        Me.MPubKey = Value
      End Set
    End Property

    Private MIsDisposed As Boolean = True
    Public Property IsDisposed() As Boolean
      Get
        Return Me.MIsDisposed
      End Get
      Set
        Me.MIsDisposed = Value
      End Set
    End Property

    Private MInfo As SlhDsaKeyPairInfo = Nothing
    Public Property Info() As SlhDsaKeyPairInfo
      Get
        Return Me.MInfo
      End Get
      Private Set
        Me.MInfo = Value
      End Set
    End Property

    Private MParameter As SlhDsaParameters = Nothing
    Public Property Parameter() As SlhDsaParameters
      Get
        Return Me.MParameter
      End Get
      Private Set
        Me.MParameter = Value
      End Set
    End Property


    Private Rand As New SecureRandom

    Public Sub New()
      Me.New(ToRngParameter())
    End Sub

    Public Sub New(parameter As SlhDsaParameters)
      Me.Parameter = parameter

      Dim generator = New SlhDsaKeyGenerationParameters(Me.Rand, parameter)

      Dim keypair_generator = New SlhDsaKeyPairGenerator()
      keypair_generator.Init(generator)

      Dim keypair = keypair_generator.GenerateKeyPair()
      Me.PubKey = DirectCast(keypair.[Public], SlhDsaPublicKeyParameters).GetEncoded()
      Dim privkey = DirectCast(keypair.[Private], SlhDsaPrivateKeyParameters).GetEncoded()
      Me.Info = New SlhDsaKeyPairInfo(Me.PubKey, privkey, parameter)

      Array.Clear(privkey)
      Me.IsDisposed = False
    End Sub

    Public Sub Clear()

      If Me.MIsDisposed Then Return

      Me.Info.Dispose()
      If Me.PubKey IsNot Nothing Then
        Array.Clear(Me.PubKey)
      End If

      Me.Rand = Nothing
      Me.Info = Nothing
      Me.PubKey = Nothing
      Me.Parameter = SlhDsaParameters.slh_dsa_sha2_128s
    End Sub

    Public Function ToPubKey() As SlhDsaPublicKeyParameters
      Return SlhDsaPublicKeyParameters.FromEncoding(Me.Parameter, Me.PubKey)
    End Function

    Public Function Sign(message As Byte()) As (Sign As Byte(), Pubkey As Byte())
      'Sign Message-Data 

      'ML-DSA Parameter 
      Dim parameter = Me.Info.ToParameter()

      Dim signer = New SlhDsaSigner(parameter, True)
      Dim privkey = SlhDsaPrivateKeyParameters.FromEncoding(parameter, Me.Info.ToPrivKey().ToBytes())
      signer.Init(True, privkey)
      signer.BlockUpdate(message, 0, message.Length)

      Return (signer.GenerateSignature(), Me.PubKey)
    End Function

    Public Shared Function Verify(
    pubkeybob As Byte(),
    signature As Byte(),
    message As Byte(),
    Parameter As SlhDsaParameters) As Boolean

      'Verify Signature
      Dim signer = New SlhDsaSigner(Parameter, True)
      Dim PubKey = SlhDsaPublicKeyParameters.FromEncoding(
        Parameter, pubkeybob)
      signer.Init(False, PubKey)
      signer.BlockUpdate(message, 0, message.Length)
      Return signer.VerifySignature(signature)
    End Function

    Private Shared Function ToRngParameter() As SlhDsaParameters
      Dim rand = New SecureRandom()
      Dim parameters = BcPqcServices.ToSLHDsaParameters()
      Dim idx = rand.[Next](parameters.Length)
      Return parameters(idx)
    End Function

    Public Shared Function ToPubKey(
      pubkey As Byte(), parameter As SlhDsaParameters) As SlhDsaPublicKeyParameters
      Return SlhDsaPublicKeyParameters.FromEncoding(parameter, pubkey)
    End Function

    Protected Overridable Sub Dispose(disposing As Boolean)
      If Not Me.IsDisposed Then
        If disposing Then Me.Clear()
        Me.IsDisposed = True
      End If
    End Sub

    Protected Overrides Sub Finalize()
      Me.Dispose(False)
    End Sub

    Public Sub Dispose() Implements IDisposable.Dispose
      Me.Dispose(True)
      GC.SuppressFinalize(Me)
    End Sub
  End Class
End Namespace
